oauth_session.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2012 Jimmy Zelinskie. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package reddit implements an abstraction for the reddit.com API.
+package geddit
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// OAuthSession represents an OAuth session with reddit.com --
+// all authenticated API calls are methods bound to this type.
+type OAuthSession struct {
+	username     string
+	password     string
+	clientID     string
+	clientSecret string
+	accessToken  string
+	tokenType    string
+	expiresIn    int
+	scope        string
+	useragent    string
+}
+
+// NewLoginSession creates a new session for those who want to log into a
+// reddit account via OAuth.
+func NewOAuthSession(username, password, useragent, clientID, clientSecret string) (*OAuthSession, error) {
+	session := &OAuthSession{
+		username:     username,
+		password:     password,
+		clientID:     clientID,
+		clientSecret: clientSecret,
+		useragent:    useragent,
+	}
+
+	err := session.newToken(&url.Values{
+		"username":   {username},
+		"password":   {password},
+		"grant_type": {"password"},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return session, nil
+}
+
+func (s *OAuthSession) newToken(postValues *url.Values) error {
```

Comment:
This was originally broken up into a separate method with the intent of it being used by whatever does re-authorization of tokens. Or however we decide on implementing implicit grant authorization.